lib/cn/cn_tree.c

```diff
@@ -1975,67 +1988,91 @@ cn_subspill_apply(struct subspill *ss)
+     * via the edge key from the previous iteration.
+     */
     while (1) {
-        struct cn_tree_node *node;
-        struct route_node *rtnext;
-        void *lock;
         struct kvset_list_entry *le;
+        struct cn_tree_node *node;
+        struct route_node *rtn;
         uint64_t node_dgen;
-        unsigned char ekey[HSE_KVS_KEY_LEN_MAX];
-        uint eklen;
+        char *wmesg;
+        void *lock;
 
         rmlock_rlock(&tree->ct_lock, &lock);
-        rtnext = rtn ? route_node_next(rtn) : route_map_first_node(tree->ct_route_map);
-        if (!rtnext) {
+        if (eklen > 0) {
+            rtn = route_map_lookup(tree->ct_route_map, ekey, eklen);
+            if (rtn) {
+                uint8_t rbuf[HSE_KVS_KEY_LEN_MAX];
+                uint rlen;
+
+                route_node_keycpy(rtn, rbuf, sizeof(rbuf), &rlen);
+                if (rlen == eklen && 0 == memcmp(rbuf, ekey, eklen))
+                    rtn = route_node_next(rtn);
+                else
+                    ev_warn(1);
+            }
+        } else {
+            rtn = route_map_first_node(tree->ct_route_map);
+        }
+
+        if (!rtn) {
             rmlock_runlock(lock);
             break;
         }
 
-        node = route_node_tnode(rtnext);
+        node = route_node_tnode(rtn);
+        if (!node->tn_route_node)
+            abort();
 
-        mutex_lock(&node->tn_ss_lock);
-        if (node->tn_ss_splitting) {
+        mutex_lock(&tree->ct_ss_lock);
+        if (node->tn_ss_splitting || node->tn_ss_joining) {
             rmlock_runlock(lock);
 
-            /* TODO: For join we'll need a refcnt on the node (to ensure that
-             * here and elsewhere the node doesn't disappear out from under us).
-             */
+            wmesg = node->tn_ss_splitting ? "spltwait" : "joinwait";
+
             atomic_inc(&tree->ct_rspill_slp);
-            cv_wait(&node->tn_ss_cv, &node->tn_ss_lock, "spltwait");
+            cv_wait(&tree->ct_ss_cv, &tree->ct_ss_lock, wmesg);
             atomic_dec(&tree->ct_rspill_slp);
 
-            mutex_unlock(&node->tn_ss_lock);
+            mutex_unlock(&tree->ct_ss_lock);
             continue;
         }
-        mutex_unlock(&node->tn_ss_lock);
 
         /* Incrementing tn_ss_spilling while holding the tree lock is sufficient
-         * to keep the node pinned while we're spilling into it.
+         * to keep both the tree and route nodes pinned across the subspill.
          */
         spillingp = &node->tn_ss_spilling;
         atomic_inc_acq(spillingp);
+        mutex_unlock(&tree->ct_ss_lock);
 
-        /* TODO: For join, rtn could become invalid after tn_ss_spilling drops
-         * to zero, so we'll likely need to use edge keys to find the next node.
+        /* rtn can't be trusted after we drop the tree lock.
          */
-        rtn = rtnext;
+        route_node_keycpy(rtn, ekey, sizeof(ekey), &eklen);
+        rtn = NULL;
```

Comment:
Using the edge keys to traverse the route map negated the need to introduce a ref count on the node and/or route map.